src/btree/bt_curnext.c

```diff
@@ -951,9 +952,17 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
          * eg: Initial data set : (1,2,3,...10) insert key 11 in a prepare transaction. loop on next
          * will return 1,2,3...10 and subsequent call to next will return a prepare conflict. Now if
          * we call prev key 10 will be returned which will be same as earlier returned key.
+         *
+         * Additionally, reset the cursor check when we are using read uncommitted isolation mode
+         * and cross a page boundary. It's possible to see out-of-order keys when the earlier
+         * returned key is removed and new keys are inserted at the start of the page.
          */
-        if (!F_ISSET(cbt, WT_CBT_ITERATE_RETRY_PREV))
+        if (!F_ISSET(cbt, WT_CBT_ITERATE_RETRY_PREV)) {
+            if (session->txn->isolation == WT_ISO_READ_UNCOMMITTED && newpage) {
```

Comment:
I think it's possible to also hit this issue with `WT_ISO_READ_COMMITTED` cursors as it's only in snapshot isolation that the cursor sees an unchanging state of the btree. What do you think about changing this to `session->txn->isolation != WT_ISO_SNAPSHOT` instead?

Reply:
{{WT_ISO_READ_COMMITTED}} also has a snapshot, we only refresh that snapshot in search. Therefore, I don't think we can see the newly inserted keys under {{WT_ISO_READ_COMMITTED}}.